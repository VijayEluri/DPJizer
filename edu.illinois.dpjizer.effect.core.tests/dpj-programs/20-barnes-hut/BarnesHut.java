/**
 * This file is licensed under the University of Illinois/NCSA Open Source License. See LICENSE.TXT for details.
 */
/**
 * Driver class for the Barnes Hut n-body simulation
 * @author Robert L. Bocchino Jr.
 * @author Rakesh Komuravelli
 */

import java.io.BufferedOutputStream;
import java.io.File;
import java.io.FileOutputStream;
import java.io.PrintStream;
import java.text.DecimalFormat;
import java.util.concurrent.CyclicBarrier;

public class BarnesHut {

	/**
	 * Number of bodies in the simulation
	 */
	private final int nbody;

	/**
	 * The geometric tree representation of the bodies
	 */
	private final Tree tree = new Tree();

	/**
	 * Constructor
	 */
	public BarnesHut(int nbody) {
		this.nbody  = nbody;
		this.tree.N = 1;
	}

	/**
	 * Constructor
	 * @param nbody Number of bodies
	 * @param nproc Number of threads
	 * @param flag Print debug info
	 */
	public BarnesHut(int nbody, int nproc, boolean flag) {
		this.nbody  = nbody;
		this.tree.N = nproc;
		this.tree.printBodies = flag;
	}

	/** print usage */
	static void printUsage() {
		System.out.print("Usage:\n");
		System.out.print("./barnes <NBODY> [N] [printOutput]\n");
		System.out.print("    where,\n");
		System.out.print("          NBODY is the number of bodies to be simulated\n");
		System.out.print("          N is the number of processors/threads\n");
		System.out.print("          third argument is for testing purposes, enter any argument to emit output, say, \"true\"\n");
		System.exit(1);
	}

	/**
	 * Program main method
	 */
	public static void main(String[] args) throws Exception {
		// Deal with args
		int nbody = 100000;
		int nproc = 1;
		boolean emitBodies = false;
		if(args.length < 1 || args.length > 3)
			printUsage();
		if(args.length == 1)
			nbody = Integer.parseInt(args[0]);
		if(args.length == 2)
		{
			nbody = Integer.parseInt(args[0]);
			nproc = Integer.parseInt(args[1]);
		}
		if(args.length == 3)
		{
			nbody = Integer.parseInt(args[0]);
			nproc = Integer.parseInt(args[1]);
			emitBodies = true;
		}

		// Create new BH object
		BarnesHut bh = new BarnesHut(nbody, nproc, emitBodies);

		// Initialize the system
		bh.initSystem(nbody);

		// Do the simulation
		bh.doSimulation();
	}

	/**
	 * Initialize the system: Create nbody bodies with random mass and
	 * position.
	 *
	 * @param nbody  Number of bodies in the simulation
	 */
	public void initSystem(int nbody) throws Exception {
		// Accumulated center of mass
		Vector cmr = new Vector();
		// Accumulated velocity
		Vector cmv = new Vector();

		//emit the input.txt which is input to splash code
		/*File f = new File(nbody+"_input.txt");
        FileOutputStream fout = new FileOutputStream(f);
        PrintStream pStr = new PrintStream(new BufferedOutputStream(fout));
        pStr.println(nbody);
        pStr.println(Constants.NDIM);
        pStr.println("0");*/

		// Fill in the tree
		tree.rmin.SETVS(-2.0);
		tree.rsize = -2.0 * -2.0;  // t->rmin.elts[0];
		tree.bodies = new Body<[i]>[nbody]#i;

		// Create an array of empty bodies
		for (int i = 0; i < nbody; ++i) {
			final int j = i;
			tree.bodies[j] = new Body<[j]>();
		}

		// Fill in the bodies, accumulating total mass and velocity.
		// For some reason we are creating 32 distinct groups of
		// bodies, each with its own "seed factor."
		for (int i = 0; i < 32; i++) {
			uniformTestdata(i, cmr, cmv);
		}

		// Normalize coordinates so average pos and vel are 0
		cmr.DIVVS(cmr, (double) nbody);
		cmv.DIVVS(cmv, (double) nbody);
		for (int i = 0; i < tree.bodies.length; ++i) {
			final int j = i;
			Body<[j]> p = tree.bodies[j];
			p.pos.SUBV(p.pos, cmr); 
			p.vel.SUBV(p.vel, cmv);
			p.index = i;
			//            p.cost = 1;
		}

		/*        //emit body masses to input file
        for(int i = 0; i < tree.bodies.length; i++)
        {
            Body p = tree.bodies[i];
            //pStr.println("0.25");
            DecimalFormat form = new DecimalFormat("#.000000");
            pStr.println(form.format(p.mass));
            //pStr.println((float)1.0/(float)(nbody/32.0));
        }
        for(int i = 0; i < tree.bodies.length; i++)
        {
            Body p = tree.bodies[i];
            for(int j = 0; j < Constants.NDIM; j++)
            {
                pStr.print(p.pos.elts[j]);
                pStr.print(" ");
            }
            pStr.println("");
        }
        for(int i = 0; i < tree.bodies.length; i++)
        {
            Body p = tree.bodies[i];
            for(int j = 0; j < Constants.NDIM; j++)
            {
                pStr.print(p.vel.elts[j]);
                pStr.print(" ");
            }
            pStr.println("");
        }
        pStr.close();*/

		//calculate bounding box once instead of expanding it everytime
		tree.setRsize();
	}

	/**
	 * Carry out the simulation
	 */
	public void doSimulation() throws InterruptedException {
		double tnow;
		double tout;
		int i, nsteps;

		/* Go through sequence of iterations */
		tnow = 0.0;
		i = 0;
		nsteps = Constants.NSTEPS;
		assert(Util.chatting("About to perform %d iters from %f to %f by %f\n",
				nsteps,tnow,Constants.tstop,Constants.dtime));

		tree.count = 0;
		long start = System.nanoTime();

		//create threads and barrier
		/*        tree.barrier = new CyclicBarrier(tree.N);
        tree.barMakeTree = new CyclicBarrier(tree.N);
        tree.barComputeGrav = new CyclicBarrier(tree.N);
        tree.barPosUpdate = new CyclicBarrier(tree.N);
        Thread th[] = new Thread[tree.N - 1];
        for(int n = 0; n < (tree.N - 1); n++)
        {
            th[n] = new Thread(new SlaveStart(n+1, tree, tnow));
            th[n].start();
        }
		 */
		i = 0;
		while ((tnow < Constants.tstop + 0.1*Constants.dtime) && (i < Constants.NSTEPS)) {
			tree.stepsystem(0, i); 
			tnow = tnow + Constants.dtime;
			assert(Util.chatting("tnow = %f sp = 0x%x\n", tnow, 0));
			i++;
		}

		/*       for(int n = 0; n < (tree.N - 1); n++)
        {
            th[n].join();
        }
		 */
		long end = System.nanoTime();
		if(!tree.printBodies)
		{
			System.out.println("Overall time taken for force calculation: " + tree.count);
			System.out.print("Overall time taken for entire program: ");
			System.out.println((end-start)/1000000000.0);
		}
	}

	/**
	 * Create uniform test data for a segment of tree.bodies.
	 *
	 * @param nbodyx      Number of bodies to fill in starting at nbodyx *
	 *                    segmentNum
	 * @param segmentNum  The number of this segment
	 * @param cmr         Accumulated center of mass
	 * @param cmv         Accumulated velocity
	 */
	private void uniformTestdata(int segmentNum, Vector cmr, Vector cmv) {
		double rsc, vsc, r, v, x, y;
		Body<*> p;
		int i;
		int seedfactor = segmentNum+1;
		double temp, t1;
		double seed = 123.0 * (double) seedfactor;
		int k;
		double rsq, rsc1;
		double rad;
		double coeff;
		int nbodyx = nbody/32; 
		double rockmass = 1.0 / (nbody/32.0);
		int start = nbodyx * segmentNum;

		rsc = 3 * Constants.PI / 16;	        /* set length scale factor  */
		vsc = Math.sqrt(1.0 / rsc);		/* and recip. speed scale   */

		for (i = 0; i < nbodyx; i++) {	        /* loop over particles      */
			/* fetch body from previously created array */
			p = tree.bodies[start+i]; 
			//p.mass = 1.0 / nbodyx;			/*   set masses equal       */
			p.mass = rockmass;			/*   set masses equal       */
			seed = Util.rand(seed);
			t1 = Util.xrand(0.0, Constants.MFRAC, seed);
			temp = Math.pow(t1,	                        /*   pick r in struct units */
					-2.0/3.0) - 1;
			r = 1 / Math.sqrt(temp);

			coeff = 4.0; /* exp(log(nbodyx/DENSITY)/3.0); */
			for (k=0; k < Constants.NDIM; k++) {
				seed = Util.rand(seed);
				r = Util.xrand(0.0, Constants.MFRAC, seed);
				p.pos.elts[k] = coeff*r;
			}

			cmr.ADDV(cmr, p.pos);		        /*   add to running sum     */
			do {					/*   select from fn g(x)    */
				seed = Util.rand(seed);
				x = Util.xrand(0.0, 1.0, seed);   	/*     for x in range 0:1   */
				seed = Util.rand(seed);
				y = Util.xrand(0.0, 0.1, seed);  	/*     max of g(x) is 0.092 */
			} while (y > x*x * Math.pow(1 - x*x, 3.5));	/*   using von Neumann tech */
			v = Math.sqrt(2.0) * 
			x / Math.pow(1 + r*r, 0.25);	        /*   find v in struct units */

			rad = vsc*v;                                /*   pick scaled velocity   */

			do {					/* pick point in NDIM-space */
				for (k = 0; k < Constants.NDIM; k++) {	/* loop over dimensions   */
					seed = Util.rand(seed);
					p.vel.elts[k] = 
						Util.xrand(-1.0, 1.0, seed);	/* pick from unit cube  */
				}
				rsq = p.vel.DOTVP(p.vel);		/*   compute radius squared */
			} while (rsq > 1.0);                	/* reject if outside sphere */
			rsc1 = rad / Math.sqrt(rsq);		/* compute scaling factor   */
			p.vel.MULVS(p.vel, rsc1);		        /* rescale to radius given  */
			cmv.ADDV(cmv, p.vel);	      	        /*   add to running sum     */
		}
	}


}
