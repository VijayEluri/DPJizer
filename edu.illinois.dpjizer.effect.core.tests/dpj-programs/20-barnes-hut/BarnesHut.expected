BarnesHut.BarnesHut(int) writes Root invokes java.lang.Object.Object()
BarnesHut.BarnesHut(int,int,boolean) writes Root invokes java.lang.Object.Object()
BarnesHut.doSimulation() writes Root invokes Tree.stepsystem(int,int) where { this -> Root : * }, Util.chatting(java.lang.String,java.lang.Object...) where { this -> Root : * }, java.io.PrintStream.print(java.lang.String) where { this -> Root : * }, java.io.PrintStream.println(double) where { this -> Root : * }, java.io.PrintStream.println(java.lang.String) where { this -> Root : * }, java.lang.System.nanoTime() where { this -> Root : * }
BarnesHut.initSystem(int) reads Root : Node.r writes Root invokes BarnesHut.uniformTestdata(int,Vector<Root>,Vector<Root>), Tree.setRsize() where { this -> Root : * }, Vector.<>DIVVS(Vector<Ru>,double) where { this -> Root : *, R -> Root, Ru -> Root }, Vector.<>SUBV(Vector<Ru>,Vector<Rw>) where { this -> Root : Node.r : *, R -> Root : Node.r, Ru -> Root : Node.r, Rw -> Root }, Vector.<>SUBV(Vector<Ru>,Vector<Rw>) where { this -> Root : [?] : *, R -> Root : [?], Ru -> Root : [?], Rw -> Root }, Vector.SETVS(double) where { this -> Root : *, R -> Root }
BarnesHut.main(java.lang.String[]) reads Root invokes BarnesHut.doSimulation() where { this -> Root : * }, BarnesHut.initSystem(int) where { this -> Root : * }, BarnesHut.printUsage(), java.lang.Integer.parseInt(java.lang.String) where { this -> Root : * }
BarnesHut.printUsage() invokes java.io.PrintStream.print(java.lang.String) where { this -> Root : * }, java.lang.System.exit(int) where { this -> Root : * }
BarnesHut.uniformTestdata(int,Vector<Root>,Vector<Root>) writes Root : * invokes Util.rand(double) where { this -> Root : * }, Util.xrand(double,double,double) where { this -> Root : * }, Vector.<>ADDV(Vector<Ru>,Vector<Rw>) where { this -> Root : *, R -> Root, Ru -> Root, Rw -> Root : * }, Vector.<>DOTVP(Vector<Ru>) where { this -> Root : *, R -> Root : *, Ru -> Root : * }, Vector.<>MULVS(Vector<Ru>,double) where { this -> Root : *, R -> Root : *, Ru -> Root : * }, java.lang.Math.pow(double,double) where { this -> Root : * }, java.lang.Math.sqrt(double) where { this -> Root : * }
Body.<>subdivp(Node,double,double,HGStruct<R>) overrides Node.<>subdivp(Node,double,double,HGStruct<R>) where { R -> R }
Body.Body() invokes Node.Node()
Body.Body(Body<Root : *>) reads Root : * writes R, Root invokes Node.Node(Node), Vector.<>SETV(Vector<Ru>) where { this -> R : *, R -> R, Ru -> Root : * }
Body.gravsub(Node,HGStruct<R>) reads Root : Node.r writes R invokes Vector.<>ADDV(Vector<Ru>,Vector<Rw>) where { this -> R : *, R -> R, Ru -> R, Rw -> R }, Vector.<>DOTVP(Vector<Ru>) where { this -> R : *, R -> R, Ru -> R }, Vector.<>MULVS(Vector<Ru>,double) where { this -> R : *, R -> R, Ru -> R }, Vector.<>SUBV(Vector<Ru>,Vector<Rw>) where { this -> R : *, R -> R, Ru -> Root : Node.r, Rw -> R }, java.lang.Math.sqrt(double) where { this -> Root : * }
Body.hackcofm() reads Root : Node.r overrides Node.hackcofm()
Body.hackgrav(HGStruct<R>,double,Node) writes R invokes Body.walksub(Node,double,double,HGStruct<R>,int) where { R -> R }, Vector.<>SETV(Vector<Ru>) where { this -> R : *, R -> R, Ru -> R }
Body.toString() reads Root : Node.r invokes java.lang.StringBuffer.append(double) where { this -> Root : * }, java.lang.StringBuffer.append(java.lang.Object) where { this -> Root : * }, java.lang.StringBuffer.append(java.lang.String) where { this -> Root : * }, java.lang.StringBuffer.toString() where { this -> Root : * } overrides java.lang.Object.toString()
Body.walksub(Node,double,double,HGStruct<R>,int) reads R, Root : Node.r invokes Body.gravsub(Node,HGStruct<R>) where { R -> R }, Body.walksub(Node,double,double,HGStruct<R>,int) where { R -> R }, Node.<>subdivp(Node,double,double,HGStruct<R>) where { this -> Root : *, R -> R }
Cell.<>subdivp(Node,double,double,HGStruct<R>) reads R, Root : Node.r invokes Vector.<>DOTVP(Vector<Ru>) where { this -> R : *, R -> R, Ru -> R }, Vector.<>SUBV(Vector<Ru>,Vector<Rw>) where { this -> R : *, R -> R, Ru -> Root : Node.r, Rw -> R } overrides Node.<>subdivp(Node,double,double,HGStruct<R>) where { R -> R }
Cell.Cell() invokes Node.Node()
Cell.hackcofm() writes Root, Root : Node.r invokes Node.hackcofm() where { this -> Root : * }, Vector.<>ADDV(Vector<Ru>,Vector<Rw>) where { this -> Root : *, R -> Root, Ru -> Root, Rw -> Root }, Vector.<>DIVVS(Vector<Ru>,double) where { this -> Root : Node.r : *, R -> Root : Node.r, Ru -> Root }, Vector.<>MULVS(Vector<Ru>,double) where { this -> Root : *, R -> Root, Ru -> Root : Node.r } overrides Node.hackcofm()
Constants.Constants() invokes java.lang.Object.Object()
HGStruct.HGStruct() writes R invokes java.lang.Object.Object()
Node.<>subdivp(Node,double,double,HGStruct<R>) pure
Node.Node() invokes java.lang.Object.Object()
Node.Node(Node) writes Root : Node.r invokes Vector.<>SETV(Vector<Ru>) where { this -> Root : Node.r : *, R -> Root : Node.r, Ru -> Root : Node.r }, java.lang.Object.Object()
Node.hackcofm() pure
SlaveStart.SlaveStart(int,Tree,double) writes Root invokes java.lang.Object.Object()
SlaveStart.run() writes Root invokes Tree.stepsystem(int,int) where { this -> Root : * }, Util.chatting(java.lang.String,java.lang.Object...) where { this -> Root : * } overrides java.lang.Runnable.run()
Tree.Tree() invokes java.lang.Object.Object()
Tree.computegrav(int,int) reads Root, Root : Node.r writes Root : [?] invokes Body.hackgrav(HGStruct<R>,double,Node) where { this -> Root : [?] : *, R -> Root : [?] }, Vector.<>ADDV(Vector<Ru>,Vector<Rw>) where { this -> Root : [?] : *, R -> Root : [?], Ru -> Root : [?], Rw -> Root : [?] }, Vector.<>MULVS(Vector<Ru>,double) where { this -> Root : [?] : *, R -> Root : [?], Ru -> Root : [?] }, Vector.<>SETV(Vector<Ru>) where { this -> Root : [?] : *, R -> Root : [?], Ru -> Root : Node.r }, Vector.<>SETV(Vector<Ru>) where { this -> Root : [?] : *, R -> Root : [?], Ru -> Root : [?] }, Vector.<>SUBV(Vector<Ru>,Vector<Rw>) where { this -> Root : [?] : *, R -> Root : [?], Ru -> Root : [?], Rw -> Root : [?] }, Vector.CLRV() where { this -> Root : [?] : *, R -> Root : [?] }
Tree.intcoord(Body<Root : *>) reads Root : Node.r writes Root invokes Vector.<>SETV(Vector<Ru>) where { this -> Root : *, R -> Root, Ru -> Root : Node.r }, java.lang.Math.floor(double) where { this -> Root : * }
Tree.loadtree(Body<Root : *>,int[],Node,int,int) writes Root : Node.r invokes Tree.intcoord(Body<Root : *>), Tree.loadtree(Body<Root : *>,int[],Node,int,int), Tree.subindex(int[],int)
Tree.maketree(int) reads Root : Node.r writes Root invokes Node.hackcofm() where { this -> Root : * }, Tree.intcoord(Body<Root : *>), Tree.loadtree(Body<Root : *>,int[],Node,int,int), Tree.reOrderBodies(Node,int), Util.chatting(java.lang.String,java.lang.Object...) where { this -> Root : * }, java.io.PrintStream.print(java.lang.String) where { this -> Root : * }, java.io.PrintStream.printf(java.lang.String,java.lang.Object...) where { this -> Root : * }, java.io.PrintStream.println(java.lang.String) where { this -> Root : * }
Tree.reOrderBodies(Node,int) writes Root, Root : Node.r invokes Tree.reOrderBodies(Node,int)
Tree.setRsize() reads Root : Node.r writes Root invokes Vector.<>ADDVS(Vector<Ru>,double) where { this -> Root : *, R -> Root, Ru -> Root }, Vector.<>SUBV(Vector<Ru>,Vector<Rw>) where { this -> Root : *, R -> Root, Ru -> Root, Rw -> Root }, Vector.SETVS(double) where { this -> Root : *, R -> Root }
Tree.stepsystem(int,int) writes Root invokes Tree.computegrav(int,int), Tree.maketree(int), Tree.setRsize(), Tree.vp(int), java.io.PrintStream.println(java.lang.String) where { this -> Root : * }, java.lang.System.nanoTime() where { this -> Root : * }
Tree.subindex(int[],int) reads Root
Tree.vp(int) reads Root, Root : Node.r invokes Vector.<>ADDV(Vector<Ru>,Vector<Rw>) where { this -> Root : *, R -> Root, Ru -> Root : [?], Rw -> Root }, Vector.<>ADDV(Vector<Ru>,Vector<Rw>) where { this -> Root : Node.r : *, R -> Root : Node.r, Ru -> Root : Node.r, Rw -> Root }, Vector.<>ADDV(Vector<Ru>,Vector<Rw>) where { this -> Root : [?] : *, R -> Root : [?], Ru -> Root, Rw -> Root }, Vector.<>MULVS(Vector<Ru>,double) where { this -> Root : *, R -> Root, Ru -> Root : [?] }, Vector.<>MULVS(Vector<Ru>,double) where { this -> Root : *, R -> Root, Ru -> Root }, java.io.PrintStream.println(java.lang.String) where { this -> Root : * }, java.lang.System.nanoTime() where { this -> Root : * }
Util.Util() invokes java.lang.Object.Object()
Util.chatting(java.lang.String,java.lang.Object...) invokes java.util.Formatter.format(java.lang.String,java.lang.Object...) where { this -> Root : * }
Util.printResults(java.lang.String) pure
Util.rand(double) invokes java.lang.Math.floor(double) where { this -> Root : * }
Util.xrand(double,double,double) pure
Vector.<>ADDV(Vector<Ru>,Vector<Rw>) reads Ru, Rw writes R
Vector.<>ADDVS(Vector<Ru>,double) reads Ru writes R
Vector.<>CROSSVP(Vector<Ru>,Vector<Rw>) reads Ru, Rw writes R
Vector.<>DISTV(double,Vector<Ru>) reads R, Ru invokes java.lang.Math.sqrt(double) where { this -> Root : * }
Vector.<>DIVVS(Vector<Ru>,double) reads Ru writes R
Vector.<>DOTVP(Vector<Ru>) reads R, Ru
Vector.<>EQUAL(Vector<Rv>) reads R, Rv
Vector.<>INCADDV(Vector<Ru>) reads Ru writes R
Vector.<>INCSUBV(Vector<Ru>) reads Ru writes R
Vector.<>MULVS(Vector<Ru>,double) reads Ru writes R
Vector.<>SETV(Vector<Ru>) reads Ru writes R
Vector.<>SUBV(Vector<Ru>,Vector<Rw>) reads Ru, Rw writes R
Vector.ABSV(double) reads R invokes java.lang.Math.sqrt(double) where { this -> Root : * }
Vector.CLRV() writes R
Vector.INCDIVVS(double) writes R
Vector.INCMULVS(double) writes R
Vector.SETVS(double) writes R
Vector.UNITV(int) writes R
Vector.Vector() invokes java.lang.Object.Object()
Vector.toString() reads R invokes java.lang.StringBuffer.append(java.lang.String) where { this -> Root : * }, java.lang.StringBuffer.toString() where { this -> Root : * }, java.util.Formatter.format(java.lang.String,java.lang.Object...) where { this -> Root : * } overrides java.lang.Object.toString()
